Sprint 12

tup1k opened the pull request.

tup1k committed:
Добавлен класс загрузки фото

tup1k committed:
Добавлена часть метода загрузки фоток без разделения массива на отдельные фото

tup1k committed:
Добавили загрузку параметров фото постранично

tup1k committed:
Откорректирован метод загрузки данных фото - появилась возможность загружать новые страницы

tup1k committed:
Добавлена загрузка фоток в ленту

tup1k committed:
Добавлено открытие картинки на большой экран и выход из аккаунта. Откорректирован шеринг.

tup1k committed:
Попытка добавить лайки

tup1k committed:
Реализована система установки лайков. Требуется проверка на стабильность.

tup1k committed:
Доработан функционал лайков и исправлена скорость загрузки фото.

tup1k committed:
Откорректирована система подгрузки ячеек. Выполнена доработка для соответствия требованиям ТЗ/ требованиям 12 спринта.

Helvetii951 commented on ImageFeed/AppDelegate.swift:
```diff
@@ -14,6 +15,9 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         // Override point for customization after application launch.
+        ProgressHUD.animationType = .activityIndicator
+        ProgressHUD.colorHUD = .black
+        ProgressHUD.colorAnimation = .lightGray
```
Лучше выносить подобные настройки в отдельную функцию. Иначе [application(_:didFinishLaunchingWithOptions:)](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622921-application) может стать громоздкой.

Helvetii951 commented on ImageFeed/Controllers/ImagesList/ImagesListCell.swift:
```diff
@@ -7,16 +7,20 @@
 final class ImagesListCell: UITableViewCell {
     static let reuseIdentifier = "ImagesListCell" // Создаем идентификатор ячейки
     private let gradientLayer = CAGradientLayer() // Создаем слой градиента
+    private let isLiked: Bool = false // Параметр определяющий наличие лайка у фото
+    weak var delegate: ImagesListCellDelegate?
```
[Документация от Apple](https://developer.apple.com/documentation/xcode/writing-symbol-documentation-in-your-source-files), если хочешь комментировать, как про :)

Helvetii951 commented on ImageFeed/Controllers/ImagesList/ImagesListViewController.swift:
```diff
@@ -7,43 +7,79 @@
-        //formatter.locale = Locale(identifier: "ru") // Русскоязычная дата
+        formatter.locale = Locale(identifier: "ru") // Русскоязычная дата
         return formatter
     }()
```
Лучше не привязваться к конкретной локали. Пусть отображение даты будет отличаться от дизайна, но пользователь увидит дату в привычном для себя виде.

Helvetii951 commented on ImageFeed/ProfileLogoutService.swift:
```diff
@@ -0,0 +1,46 @@
+                cache.clearDiskCache()
+                cache.cleanExpiredMemoryCache()
+                cache.clearCache()
+    }
```
Чуть поехали отступы. Они помогают коду выглядеть аккуратнее и привычнее взгляду, так читать код намного удобнее. Чтобы поправить отступы: можно выделить весь код `cmd+A` и зажать `contol+i`.

Helvetii951 commented on ImageFeedTests/ImagesListServiceTests.swift:
```diff
@@ -0,0 +1,63 @@
+@testable import ImageFeed
+import XCTest
+
+final class ImagesListServiceTests: XCTestCase {
+    private let tokenStoragePVC = OAuth2TokenStorage()
+    
+    func testFetchPhotos() {
+        guard let token = tokenStoragePVC.token else { return }
+        let service = ImagesListService.shared
+        
+        let expectation = self.expectation(description: "Wait for Notification")
+        NotificationCenter.default.addObserver(
+            forName: ImagesListService.didChangeNotification,
+            object: nil,
+            queue: .main) { _ in
+                expectation.fulfill()
+            }
```
Писать тесты - это очень клево👍

Helvetii951 commented on ImageFeed/Controllers/ImagesList/ImagesListViewController.swift:
```diff
@@ -7,43 +7,79 @@
+            cell.dateLabel.text = dateFormatter.string(from: newImageCreationDate) // Присваиваем дату в нужном формате аутлету даты
+        } else {
+            cell.dateLabel.text = "" // Если даты нет ставим пустую строку
+        }
```
Swift - достаточно выразительный язык. Часто можно понять что происходит и без комментариев :)

Helvetii951 commented on ImageFeed/Controllers/ImagesList/ImagesListCell.swift:
```diff
@@ -30,7 +34,24 @@ final class ImagesListCell: UITableViewCell {
+        cellImage.kf.cancelDownloadTask()
+    }
+    
+    @IBAction func likeButtonPushed(_ sender: Any) {
+        delegate?.imageListCellDidTapLike(self)
+    }
```
Функцию можно сделать приватной. Рекомендую всегда писать функции и свойства с минимальной областью видимости, а затем менять ее, если нужно.

Helvetii951 commented on ImageFeed/Controllers/ProfileViewController.swift:
```diff
@@ -138,4 +145,34 @@ final class ProfileViewController: UIViewController {
+    
+    //  Функция выхода из приложения
+    @objc
+    private func tapLogOutButton() {
+        let alert = UIAlertController(title: "Внимание!", message: "Вы уверены что хотите выйти?", preferredStyle: .alert)
+        let yesButtonAction = UIAlertAction(title: "Да", style: .default, handler: { [weak self] _ in
+            guard let self = self else { return }
+            profileLogOut.logout()
+            profileExitTransit()
+            print("ВЫ ВЫШЛИ ИЗ ПРОФИЛЯ. ТОКЕН - \(myToken.token ?? "НЕ ИДЕНТИФИЦИИРУЕТСЯ")")
+        })
```
Текст алерта расходится с дизайном. Часто такое не принимают тестировщики. Лучше исправить.

tup1k replied:
Я поправил алерт. А ты не подскажешь, у тебя в замечании написано, что окно профиля без кнопки таббара - я не могу воспроизвести эту ошибку, у меня и фон черный и кнопки таббара есть. Попросил ребят запустить приложение и у них тоже есть кнопки. Но у них xcode тоже предыдущий. Может дело в нем. Я пробовал сбрасывать все настройки в симуляторе но ничего не изменилось.![image](https://github.com/user-attachments/assets/85596b4f-7d83-4326-a2b5-9f12ab628d30)

Helvetii951 replied:
Привет! Попробуй запушить все изменения перед сдачей, оставь коммент. Я его увижу, посмотрю.

tup1k replied:
Привет. Я запушил изменения в ветку 12 спринта (https://github.com/tup1k/ImageFeed/tree/sprint_12). Если будет возможность посмотри пожалуйста.

Helvetii951 replied:
Можно сдавать работу.

tup1k committed:
Выполнены следующие корректировки:
- настройка ProgressHUD вынесена в отдельный метод;
- усовершенствованы комментарии;
- удалена привязка к конкретной локализации дат;
- ряд функций, для которых существовала такая возможность, переведены в разряд приватных;
- текст алерта выхода из профиля приведен в соответствие с ТЗ (Figma).
